ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/MLSdkAsyncHttpResponseHandler.java

```diff
@@ -197,27 +136,46 @@ public void onError(Throwable t) {
                     t instanceof NullPointerException ? "NullPointerException" : t.getMessage(),
                     t
                 );
-            response(tensorOutputs);
+            response();
         }
 
         @Override
         public void onComplete() {
-            response(tensorOutputs);
+            response();
         }
     }
 
-    private void response(Map<Integer, ModelTensors> tensors) {
-        processResponse(statusCode, responseBody.toString(), parameters, tensorOutputs);
-        executionContext.getCountDownLatch().countDown();
-        // when countdown's count equals to 0 means all responses are received.
-        if (executionContext.getCountDownLatch().getCount() == 0) {
-            if (executionContext.getExceptionHolder().get() != null) {
-                actionListener.onFailure(executionContext.getExceptionHolder().get());
-                return;
+    private void response() {
+        String body = responseBody.toString();
+        if (Strings.isBlank(body)) {
+            log.error("Remote model response body is empty!");
+            actionListener.onFailure(new OpenSearchStatusException("No response from model", RestStatus.BAD_REQUEST));
+            return;
+        }
+
+        if (statusCode < HttpStatus.SC_OK || statusCode > HttpStatus.SC_MULTIPLE_CHOICES) {
+            if (containsThrottlingException) {
+                actionListener
+                    .onFailure(
+                        new RetryableException(
+                            REMOTE_SERVICE_ERROR + "The request was denied due to remote server throttling.",
+                            RestStatus.fromCode(statusCode)
+                        )
+                    );
```

Comment:
Is there any other benefits of removing exceptionHolder? I feel like this is adding a lot of manual checkups for different exceptions. For handling more other exceptions, this would needs to add more code and flags? I think this is less flexible than using exceptionHolder, even we don't need to countDown any more in the retry.

Reply:
Good point, ack